lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -452,66 +503,141 @@ def name(self) -> str:
+            return {}
+        result = dict(self.get_relation().data[self._relation.app]) or {}
+        if not self.is_main_orchestrator:
+            # Peer relations exchange secrets via peer-cluster secret
+            result |= self._charm.secrets.get_object(Scope.APP, S3_CREDENTIALS) or {}
+        return result
+
+
 class OpenSearchBackupPlugin(OpenSearchPlugin):
     """Manage backup configurations.
 
     This class must load the opensearch plugin: repository-s3 and configure it.
+
+    The plugin is responsible for managing the backup configuration, which includes relation
+    databag or only the secrets' content, as backup changes behavior depending on the juju app
+    role within the cluster.
     """
 
+    MODEL = S3RelData
+    MANDATORY_CONFS = [
+        "bucket",
+        "endpoint",
+        "region",
+        "base_path",
+        "protocol",
+        "credentials",
+    ]
+    DATA_PROVIDER = OpenSearchPluginBackupDataProvider
+
+    def __init__(self, charm):
+        """Creates the OpenSearchBackupPlugin object."""
+        super().__init__(charm)
+        self.dp = self.DATA_PROVIDER(charm)
+        self.repo_name = "default"
+
+    def requested_to_enable(self) -> bool:
+        """Returns True if the plugin is enabled."""
+        return self.dp.get_relation() is not None
+
+    @property
+    def data(self) -> BaseModel:
+        """Returns the data from the relation databag."""
+        self._relation = self.dp.get_relation()
+        try:
+            return self.MODEL.from_relation(self.dp.get_data())
+        except ValidationError:
+            return self.MODEL()
+
     @property
     def name(self) -> str:
         """Returns the name of the plugin."""
         return "repository-s3"
 
     def config(self) -> OpenSearchPluginConfig:
-        """Returns OpenSearchPluginConfig composed of configs used at plugin addition.
-
-        Format:
-        OpenSearchPluginConfig(
-            config_entries_to_add = {...},
-            config_entries_to_del = [...],
-            secret_entries_to_add = {...},
-            secret_entries_to_del = [...],
-        )
-        """
-        if not self._extra_config.get("access-key") or not self._extra_config.get("secret-key"):
+        """Returns OpenSearchPluginConfig composed of configs used at plugin configuration."""
+        conf = self.data.credentials.dict()
+        # First, let's check if credentials are set
+        if any([val is None for val in conf.values()]):
             raise OpenSearchPluginMissingConfigError(
-                "Plugin {} missing: {}".format(
+                "Plugin {} missing credentials".format(
                     self.name,
-                    [
-                        conf
-                        for conf in ["access-key", "secret-key"]
-                        if not self._extra_config.get(conf)
-                    ],
                 )
             )
 
+        if self.dp.is_main_orchestrator:
```

Comment:
this is too smart for the plugins - which is supposed to be dumb. The heavy lifting should have been on the data provider and only inject to the plugin what it needs.

Reply:
Yeah, I agree. But the thing here is that I need some "factory" method that differs between MAIN or OTHERS...